codec/src/main/java/io/netty/handler/codec/compression/ZstdDecoder.java

```diff
@@ -0,0 +1,186 @@
+    private ZstdInputStreamNoFinalizer zstdIs;
+    private ByteArrayOutputStream bos;
+    private final int maxBlockSize;
+    private ByteBuf buffer;
+
+    /**
+     * Creates a new Zstd decoder.
+     *
+     * Please note that if you use the default constructor, the MAX_BLOCK_SIZE
+     * will be used. If you want to specify MAX_BLOCK_SIZE yourself,
+     * please use {@link ZstdDecoder(int)} constructor
+     */
+    public ZstdDecoder() {
+        this(DEFAULT_MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd decoder.
+     *  @param  maxBlockSize
+     *            specifies the max block size
+     */
+    public ZstdDecoder(int maxBlockSize) {
+        this.maxBlockSize = ObjectUtil.checkPositive(maxBlockSize, "maxBlockSize");
+    }
+
+    /**
+     * Current state of stream.
+     */
+    private enum State {
+        DECOMPRESS_DATA,
+        NEED_MORE_DATA,
+        FINISHED,
+        CORRUPTED
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            while (in.isReadable()) {
+                switch (currentState) {
+                    case DECOMPRESS_DATA:
+                        decompressData(ctx, in, out);
+                        break;
+                    case NEED_MORE_DATA:
+                        buffer.writeBytes(in);
+                        decompressData(ctx, buffer, out);
+                        break;
+                    case FINISHED:
+                    case CORRUPTED:
+                        in.skipBytes(in.readableBytes());
+                        break;
+                    default:
+                        throw new IllegalStateException();
+                }
+            }
+        } catch (Exception e) {
+            currentState = State.CORRUPTED;
+            throw e;
+        }
+    }
+
+    private boolean consumeAndDecompress(ChannelHandlerContext ctx, int decompressedSize,
+                                         ByteBuf in, List<Object> out) throws IOException {
+        // Bytebuf cannot release here because ByteToMessageDecoder will release it
+        is = new ByteBufInputStream(in, false);
+        zstdIs = new ZstdInputStreamNoFinalizer(is);
+        // setContinuous to true so that ZstdInputStreamNoFinalizer.read() will
+        // return -1 when decompression is not completed
+        zstdIs.setContinuous(true);
+        bos = new ByteArrayOutputStream();
+        IOUtils.copy(zstdIs, bos);
+        byte[] decompressed = bos.toByteArray();
+        int decompressedLength = decompressed.length;
+        // Check the decompression status since we use ZstdInputStream as an indicator
+        // to determine whether the decompression is completed
+        // TODO: Use Zstd to check the compressed data when zstd library support this
+        if (decompressedLength == 0 || (decompressedSize > 0 && decompressedSize != decompressedLength)) {
+            in.readerIndex(0);
```

Comment:
I this correct ? I think we should reset the readerIndex to what it was before which might not be 0.

Reply:
Done